c++/src/kj/async.c++

```diff
@@ -290,7 +290,11 @@ kj::String TaskSet::trace() {
 }
 
 Promise<void> TaskSet::onEmpty() {
-  KJ_REQUIRE(emptyFulfiller == nullptr, "onEmpty() can only be called once at a time");
+  KJ_IF_MAYBE(fulfiller, emptyFulfiller) {
+    if (fulfiller->get()->isWaiting()) {
+      KJ_REQUIRE(emptyFulfiller == nullptr, "onEmpty() can only be called once at a time");
```

Comment:
Since we already know the fulfiller is non-null, I'd change this to:

    KJ_FAIL_REQUIRE("onEmpty() can only be called once at a time");